microscope/cameras/ximea.py

```diff
@@ -0,0 +1,161 @@
+        self._exposure_time = 0.1
+        self._triggered = False
+
+    def _fetch_data(self):
+        if self._acquiring and self._triggered:
+            self.handle.get_image(self.img)
+            self._logger.info('Sending image')
+            self._triggered = False
+            data = np.array(list(self.img.get_image_data_raw()),dtype=np.uint16).reshape(self.img.width,self.img.height)
+            return data
+
+    def abort(self):
+        self._logger.info('Disabling acquisition.')
+        if self._acquiring:
+            self._acquiring = False
+        self.handle.stop_acquisition()
+
+    def initialize(self):
+        """Initialise the camera.
+
+        Open the connection, connect properties and populate settings dict.
+        """
+        try:
+            self.handle = xiapi.Camera()
+            self.handle.open_device()
+        except:
+            raise Exception("Problem opening camera.")
+        if self.handle == None:
+            raise Exception("No camera opened.")
+        self._logger.info('Initializing.')
+        self.img=xiapi.Image()
+
+    def make_safe(self):
+        if self._acquiring:
+            self.abort()
+
+    def _on_disable(self):
+        self.abort()
+
+    def _on_enable(self):
+        self._logger.info("Preparing for acquisition.")
+        if self._acquiring:
+            self.abort()
+        self._acquiring = True
+        #actually start camera
+        self.handle.start_acquisition()
+        self._logger.info("Acquisition enabled.")
+        return True
+
+    def set_exposure_time(self, value):
+        #exposure times are set in us.
+        self.handle.set_exposure(int(value*1.0E6))
+
+    def get_exposure_time(self):
+        #exposure times are in us, so multiple by 1E-6 to get seconds.
+        return (self.handle.get_exposure()*1.0E-6) 
+
+    def get_cycle_time(self):
+        return (self.handle.get_exposure()*1.0E-6)
+
+    def _get_sensor_shape(self):
+        return (self.img.width,self.image.height)
+
+
+    def get_trigger_type(self):
+        trig=self.handle.get_trigger_source()
+        return devices.TRIGGER_SOFT
+
+    def set_trigger_type(self, trigger):
+        if (trigger == devices.TRIGGER_SOFT):
+            self.handle.set_triger_source(XI_TG_SOURCE['Xi_TRG_SOFTWARE'])
+        elif (trigger == devices.TRIGGER_BEFORE):
+            self.handle.set_triger_source(XI_TG_SOURCE['Xi_TRG_EDGE_RISING'])
+            #define digial input mode of trigger
+            self.handle.set_gpi_selector(1)
+            self.handle.set_gpi_mode(XI_GPI_TRIGGER)
+
+    def soft_trigger(self):
+        self._logger.info('Trigger received; self._acquiring is %s.'
+                          % self._acquiring)
+        if self._acquiring:
+            self._triggered = True
+
+    def _get_binning(self):
+         return (1,1)
+
+    @keep_acquiring
+    def _set_binning(self, h, v):
+        return False
+
+    def _get_roi(self):
+        return (0, 0, 512, 512)
```

Comment:
If ROI and binning are not implemented, this should return the sensor size, not 512 x 512.

Reply:
Yes, obviously a terrible hack, Shows that this code is no where near complete. Sorry.

Reply:
Fixed by f2680e33b937deee75d7225105afe017408900ec